### PR TITLE
Rename RzTypeCType to RzType, other refactoring

### PR DESCRIPTION
**Pull request by XVilka**
# DO NOT SQUASH

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

A preparation and just a better name for the C/C++ Tree-Sitter-based parser. Since Tree-Sitter produces something closer to AST (not exactly, it's actually concrete syntax tree but can be omitted for simplicity), it makes sense to name these functions as `RzType`. It also avoids unnecessary `Type` duplication.

Another change here is the consolidation of the APIs from `librz/type/type.c` and `librz/type/utype.c` since the second becomes unnecessary.

Third commit is removing the remaining direct access to types SDB from `librz/analysis/var.c`.

**Test plan**

CI is green.

**Closing issues**

Related to https://github.com/rizinorg/rizin/issues/371 and https://github.com/rizinorg/rizin/issues/275

### CODE REVIEW

File: librz/type/type.c [hunk 1]
```diff
@@ -673,3 +673,34 @@ RZ_API char *rz_type_db_get_struct_member(RzTypeDB *typedb, RZ_NONNULL const cha
 	free(members);
 	return res;
 }
+
+RZ_API RZ_OWN RzList /* RzTypeStructMember */ *rz_type_db_struct_member_list(RzTypeDB *typedb, RZ_NONNULL const char *type) {
```
Comment (thestr4ng3r) [inline librz/type/type.c]: You should be able to just use `rz_type_db_get_base_type()` instead of introducing a new function.

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
Up for debate, but I would simply call it `RzType` instead of `RzASTType`. Eventually we will want to use these types everywhere and they will get api for determining their size, etc. so it's not just for AST.

**Comment by XVilka** [comment]
@thestr4ng3r RzASTType doesn't have a concept of size or anything like that - it's just a semantic piece of C/C++ AST

**Comment by thestr4ng3r** [comment]
It does not store explicit sizes but it will still have a concrete size in a certain context. For example a pointer type has the size of an address, identifier have the size of the respective base type, etc. See also https://github.com/rizinorg/rizin/issues/371 so in the end for example variables just have an `RzType` and their size is determined by evaluating the size of this type.
Does this concept somehow clash with implementing the tree-sitter parser?

**Comment by XVilka** [comment]
@thestr4ng3r there is no size of the address at the parsing stage - it depends on the values in `RzTypeTarget` structure: https://github.com/rizinorg/rizin/blob/dev/librz/include/rz_type.h#L20
It's done so we can parse C type once, store the AST, then different values of `RzTypeTarget` can be "applied" to that AST type description without reparsing everything again. Also storing AST can help to restore the type as a string easier.

**Comment by thestr4ng3r** [comment]
Yes, that is what I mean too. These kinds of types are somewhat syntactical and get their exact meaning and memory layout from the context where they are used. But if possible we will want to use these types directly everywhere in the analysis too where it's not necessarily AST-related so `RzType` is a better fit imo. What do you think?

File: librz/type/type.c (hunk 1 above)
Comment (XVilka) [reply librz/type/type.c]: Good catch, indeed it's easier and shorter. Done.

**Review by thestr4ng3r — DISMISSED** [review]

File: librz/include/rz_type.h [hunk 2]
```diff
@@ -33,21 +33,29 @@ typedef struct rz_type_t {
...
+// "base" types that are types already having the
+// concrete size and memory layout
+// or the "AST" types that are returned from the parser
+// and don't contain the size or memory laoyout
+
+// Base types
+
 typedef struct rz_type_enum_case_t {
-	char *name;
+	const char *name;
```
Comment (thestr4ng3r) [inline librz/include/rz_type.h]: Is this really correct? Iirc the RzTypeEnumCase and others own these strings.

Comment (XVilka) [reply librz/include/rz_type.h]: Fixed.

**Review by thestr4ng3r — APPROVED** [review]

**Comment by XVilka** [comment]
@thestr4ng3r there is one failing test, but looks like new output is actually better:
```diff
[XX] db/cmd/dwarf "function info integration 2 rust"
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e asm.bytes=true
e asm.dwarf=false
aaa
pd 40 @  dbg.bubble_sort_i32
' bins/elf/dwarf_rust_bubble
-- stdout
@@ -16,19 +16,17 @@
 |           ; var int64_t var_60h @ rsp+0x60
 |           ; var int64_t var_68h @ rsp+0x68
 |           ; var Range<usize> iter @ rsp+0x70
-|           ; var int64_t var_78h @ rsp+0x78
 |           ; var int64_t var_80h @ rsp+0x80
 |           ; var int64_t var_88h @ rsp+0x88
 |           ; var &mut [i32] values @ rsp+0x90
-|           ; var int64_t var_98h @ rsp+0x98
 |           ; var usize val @ rsp+0xa0
 |           ; var usize __next @ rsp+0xa8
 |           ; var usize i @ rsp+0xb0
 |           ; arg int64_t arg1 @ rdi
 |           ; arg int64_t arg2 @ rsi
 |           0x00005270      4881ecb80000.  sub   rsp, 0xb8             ; void bubble_sort<i32>(&mut [i32] values);
-|           0x00005277      4889bc249000.  mov   qword [values], rdi   ; arg1
-|           0x0000527f      4889b4249800.  mov   qword [var_98h], rsi  ; arg2
+|           0x00005277      4889bc249000.  mov   qword [values.data_ptr], rdi ; arg1
+|           0x0000527f      4889b4249800.  mov   qword [rsp + 0x98], rsi ; arg2
 |           0x00005287      48897c2448     mov   qword [var_48h], rdi  ; arg1
 |           0x0000528c      4889742440     mov   qword [var_40h], rsi  ; arg2
 |           ; DATA XREF from core::fmt::builders::DebugTuple::finish::hfbbe40c195c334d6 @ 0x28b56
@@ -51,11 +49,11 @@
 |           0x000052d6      4889442438     mov   qword [var_38h], rax
 |           0x000052db      4889542430     mov   qword [var_30h], rdx
 |           0x000052e0      488b442438     mov   rax, qword [var_38h]
-|           0x000052e5      4889442470     mov   qword [iter], rax
+|           0x000052e5      4889442470     mov   qword [iter.start], rax
 |           0x000052ea      488b4c2430     mov   rcx, qword [var_30h]
-|           0x000052ef      48894c2478     mov   qword [var_78h], rcx
+|           0x000052ef      48894c2478     mov   qword [rsp + 0x78], rcx
 |           ; CODE XREF from rust::bubble_sort::h0777bc845caabc60 @ 0x5432
-|           0x000052f4      488d7c2470     lea   rdi, [iter]           ; int64_t arg1
+|           0x000052f4      488d7c2470     lea   rdi, [iter.start]     ; int64_t arg1
 |           0x000052f9      e812130000     call  sym core::iter::range::_<impl core::iter::traits::iterator::Iterator for core::ops::range::Range<A>>::next::h707e8283b20ce50a ; dbg.next<usize>
 |           0x000052fe      488994248800.  mov   qword [var_88h], rdx
 |           0x00005306      488984248000.  mov   qword [var_80h], rax

```

This is what I have in IDA Pro 7.6:
```
text:0000000000005270 ; =============== S U B R O U T I N E =======================================
.text:0000000000005270
.text:0000000000005270 ; Attributes: static
.text:0000000000005270
.text:0000000000005270 ; void __cdecl rust::bubble_sort::h0777bc845caabc60(_mut__i32_ values)
.text:0000000000005270 _ZN4rust11bubble_sort17h0777bc845caabc60E proc near
.text:0000000000005270                                         ; CODE XREF: rust::main::h523dcf5432fcfd88+125↓p
.text:0000000000005270
.text:0000000000005270 a               = qword ptr -0B8h
.text:0000000000005270 var_A9          = byte ptr -0A9h
.text:0000000000005270 var_A8          = qword ptr -0A8h
.text:0000000000005270 index           = qword ptr -0A0h
.text:0000000000005270 b               = qword ptr -98h
.text:0000000000005270 var_90          = qword ptr -90h
.text:0000000000005270 var_88          = qword ptr -88h
.text:0000000000005270 var_80          = qword ptr -80h
.text:0000000000005270 len             = qword ptr -78h
.text:0000000000005270 var_70          = _mut__i32_ ptr -70h
.text:0000000000005270 var_59          = byte ptr -59h
.text:0000000000005270 self            = core::ops::range::Range<usize> ptr -58h
.text:0000000000005270 var_48          = core::option::Option<usize> ptr -48h
.text:0000000000005270 var_38          = qword ptr -38h
.text:0000000000005270 var_30          = qword ptr -30h
.text:0000000000005270 var_28          = qword ptr -28h
.text:0000000000005270 var_20          = qword ptr -20h
.text:0000000000005270 var_18          = qword ptr -18h
.text:0000000000005270 var_10          = qword ptr -10h
.text:0000000000005270 var_8           = qword ptr -8
.text:0000000000005270 n               = qword ptr  58h
.text:0000000000005270 swapped         = byte ptr  67h
.text:0000000000005270 iter            = core::ops::range::Range<usize> ptr  78h
.text:0000000000005270 values          = _mut__i32_ ptr  98h
.text:0000000000005270 val             = qword ptr  0A8h
.text:0000000000005270 __next          = qword ptr  0B0h
.text:0000000000005270 i               = qword ptr  0B8h
.text:0000000000005270
.text:0000000000005270 ; __unwind {
.text:0000000000005270                 sub     rsp, 0B8h
.text:0000000000005277                 mov     [rsp+0B8h+var_28], rdi
.text:000000000000527F                 mov     [rsp+0B8h+var_20], rsi
.text:0000000000005287                 mov     [rsp+0B8h+var_70.data_ptr], rdi
.text:000000000000528C                 mov     [rsp+0B8h+len], rsi
.text:0000000000005291                 call    _ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$3len17hbaf7153778228b73E ; core::slice::_$LT$impl$u20$$u5b$T$u5d$$GT$::len::hbaf7153778228b73
.text:0000000000005296                 mov     [rsp+0B8h+var_70.length], rax
.text:000000000000529B                 mov     [rsp+0B8h+var_59], 1
.text:00000000000052A0
.text:00000000000052A0 loc_52A0:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+1D1↓j
.text:00000000000052A0                 test    [rsp+0B8h+var_59], 1
.text:00000000000052A5                 jnz     short loc_52AF
.text:00000000000052A7                 add     rsp, 0B8h
.text:00000000000052AE                 retn
.text:00000000000052AF ; ---------------------------------------------------------------------------
.text:00000000000052AF
.text:00000000000052AF loc_52AF:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+35↑j
.text:00000000000052AF                 mov     [rsp+0B8h+var_59], 0
.text:00000000000052B4                 mov     rax, [rsp+0B8h+var_70.length]
.text:00000000000052B9                 mov     [rsp+0B8h+self.start], 1
.text:00000000000052C2                 mov     [rsp+0B8h+self.end], rax
.text:00000000000052C7                 mov     rdi, [rsp+0B8h+self.start] ; self
.text:00000000000052CC                 mov     rsi, [rsp+0B8h+self.end] ; self
.text:00000000000052D1                 call    _ZN63_$LT$I$u20$as$u20$core__iter__traits__collect__IntoIterator$GT$9into_iter17h12f1e7995fb38cf5E ; _$LT$I$u20$as$u20$core..iter..traits..collect..IntoIterator$GT$::into_iter::h12f1e7995fb38cf5
.text:00000000000052D6                 mov     [rsp+0B8h+var_80], rax
.text:00000000000052DB                 mov     [rsp+0B8h+var_88], rdx
.text:00000000000052E0                 mov     rax, [rsp+0B8h+var_80]
.text:00000000000052E5                 mov     qword ptr [rsp+0B8h+var_48.gap0], rax
.text:00000000000052EA                 mov     rcx, [rsp+0B8h+var_88]
.text:00000000000052EF                 mov     qword ptr [rsp+0B8h+var_48.gap0+8], rcx
.text:00000000000052F4
.text:00000000000052F4 loc_52F4:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60:loc_5432↓j
.text:00000000000052F4                 lea     rdi, [rsp+0B8h+var_48] ; retstr
.text:00000000000052F9                 call    _ZN4core4iter5range101_$LT$impl$u20$core__iter__traits__iterator__Iterator$u20$for$u20$core__ops__range__Range$LT$A$GT$$GT$4next17h707e8283b20ce50aE ; core::iter::range::_$LT$impl$u20$core..iter..traits..iterator..Iterator$u20$for$u20$core..ops..range..Range$LT$A$GT$$GT$::next::h707e8283b20ce50a
.text:00000000000052FE                 mov     [rsp+0B8h+var_30], rdx
.text:0000000000005306                 mov     [rsp+0B8h+var_38], rax
.text:000000000000530E                 mov     rax, [rsp+0B8h+var_38]
.text:0000000000005316                 test    rax, rax
.text:0000000000005319                 jz      short loc_531F
.text:000000000000531B                 jmp     short $+2
.text:000000000000531D ; ---------------------------------------------------------------------------
.text:000000000000531D
.text:000000000000531D loc_531D:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+AB↑j
.text:000000000000531D                 jmp     short loc_5342
.text:000000000000531F ; ---------------------------------------------------------------------------
.text:000000000000531F
.text:000000000000531F loc_531F:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+A9↑j
.text:000000000000531F                 mov     rax, [rsp+0B8h+var_70.length]
.text:0000000000005324                 sub     rax, 1
.text:000000000000532A                 setb    cl
.text:000000000000532D                 test    cl, 1
.text:0000000000005330                 mov     [rsp+0B8h+var_90], rax
.text:0000000000005335                 jnz     loc_54BA
.text:000000000000533B                 jmp     loc_5437
.text:0000000000005340 ; ---------------------------------------------------------------------------
.text:0000000000005340                 ud2
.text:0000000000005342 ; ---------------------------------------------------------------------------
.text:0000000000005342
.text:0000000000005342 loc_5342:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60:loc_531D↑j
.text:0000000000005342                 mov     rax, [rsp+0B8h+var_30]
.text:000000000000534A                 mov     [rsp+0B8h+var_18], rax
.text:0000000000005352                 mov     [rsp+0B8h+var_10], rax
.text:000000000000535A                 mov     [rsp+0B8h+var_8], rax
.text:0000000000005362                 mov     rcx, rax
.text:0000000000005365                 sub     rcx, 1
.text:000000000000536C                 setb    dl
.text:000000000000536F                 test    dl, 1
.text:0000000000005372                 mov     [rsp+0B8h+b], rax
.text:0000000000005377                 mov     [rsp+0B8h+index], rcx
.text:000000000000537C                 jnz     loc_5446
.text:0000000000005382                 mov     rax, [rsp+0B8h+index]
.text:0000000000005387                 mov     rcx, [rsp+0B8h+len]
.text:000000000000538C                 cmp     rax, rcx
.text:000000000000538F                 setb    dl
.text:0000000000005392                 test    dl, 1
.text:0000000000005395                 jnz     short loc_539C
.text:0000000000005397                 jmp     loc_5464
.text:000000000000539C ; ---------------------------------------------------------------------------
.text:000000000000539C
.text:000000000000539C loc_539C:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+125↑j
.text:000000000000539C                 mov     rax, [rsp+0B8h+index]
.text:00000000000053A1                 shl     rax, 2
.text:00000000000053A5                 mov     rcx, [rsp+0B8h+var_70.data_ptr]
.text:00000000000053AA                 add     rcx, rax
.text:00000000000053AD                 mov     rax, [rsp+0B8h+b]
.text:00000000000053B2                 mov     rdx, [rsp+0B8h+len]
.text:00000000000053B7                 cmp     rax, rdx
.text:00000000000053BA                 setb    sil
.text:00000000000053BE                 test    sil, 1
.text:00000000000053C2                 mov     [rsp+0B8h+var_A8], rcx
.text:00000000000053C7                 jnz     short loc_53CE
.text:00000000000053C9                 jmp     loc_5480
.text:00000000000053CE ; ---------------------------------------------------------------------------
.text:00000000000053CE
.text:00000000000053CE loc_53CE:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+157↑j
.text:00000000000053CE                 mov     rax, [rsp+0B8h+b]
.text:00000000000053D3                 shl     rax, 2
.text:00000000000053D7                 mov     rcx, [rsp+0B8h+var_70.data_ptr]
.text:00000000000053DC                 add     rcx, rax
.text:00000000000053DF                 mov     rdi, [rsp+0B8h+var_A8] ; self
.text:00000000000053E4                 mov     rsi, rcx        ; other
.text:00000000000053E7                 call    _ZN4core3cmp5impls55_$LT$impl$u20$core__cmp__PartialOrd$u20$for$u20$i32$GT$2gt17hcb39ce7110420639E ; core::cmp::impls::_$LT$impl$u20$core..cmp..PartialOrd$u20$for$u20$i32$GT$::gt::hcb39ce7110420639
.text:00000000000053EC                 mov     [rsp+0B8h+var_A9], al
.text:00000000000053F0                 mov     al, [rsp+0B8h+var_A9]
.text:00000000000053F4                 test    al, 1
.text:00000000000053F6                 jnz     short loc_53FA
.text:00000000000053F8                 jmp     short loc_5432
.text:00000000000053FA ; ---------------------------------------------------------------------------
.text:00000000000053FA
.text:00000000000053FA loc_53FA:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+186↑j
.text:00000000000053FA                 mov     rax, [rsp+0B8h+b]
.text:00000000000053FF                 sub     rax, 1
.text:0000000000005405                 setb    cl
.text:0000000000005408                 test    cl, 1
.text:000000000000540B                 mov     [rsp+0B8h+a], rax
.text:000000000000540F                 jnz     loc_549C
.text:0000000000005415                 mov     rdi, [rsp+0B8h+var_70.data_ptr] ; self
.text:000000000000541A                 mov     rsi, [rsp+0B8h+len] ; self
.text:000000000000541F                 mov     rdx, [rsp+0B8h+a] ; a
.text:0000000000005423                 mov     rcx, [rsp+0B8h+b] ; b
.text:0000000000005428                 call    _ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$4swap17h5ca623ef598b2bb1E ; core::slice::_$LT$impl$u20$$u5b$T$u5d$$GT$::swap::h5ca623ef598b2bb1
.text:000000000000542D                 mov     [rsp+0B8h+var_59], 1
.text:0000000000005432
.text:0000000000005432 loc_5432:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+188↑j
.text:0000000000005432                 jmp     loc_52F4
.text:0000000000005437 ; ---------------------------------------------------------------------------
.text:0000000000005437
.text:0000000000005437 loc_5437:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+CB↑j
.text:0000000000005437                 mov     rax, [rsp+0B8h+var_90]
.text:000000000000543C                 mov     [rsp+0B8h+var_70.length], rax
.text:0000000000005441                 jmp     loc_52A0
.text:0000000000005446 ; ---------------------------------------------------------------------------
.text:0000000000005446
.text:0000000000005446 loc_5446:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+10C↑j
.text:0000000000005446                 lea     rdi, str_0      ; expr
.text:000000000000544D                 lea     rdx, off_382B8  ; "rust.rs"
.text:0000000000005454                 lea     rax, core__panicking__panic ; RUST NAME: core::panicking::panic
.text:000000000000545B                 mov     esi, 21h ; '!'  ; expr
.text:0000000000005460                 call    rax ; core__panicking__panic ; RUST NAME: core::panicking::panic
.text:0000000000005462 ; ---------------------------------------------------------------------------
.text:0000000000005462                 ud2
.text:0000000000005464 ; ---------------------------------------------------------------------------
.text:0000000000005464
.text:0000000000005464 loc_5464:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+127↑j
.text:0000000000005464                 lea     rdi, location   ; location
.text:000000000000546B                 lea     rax, core__panicking__panic_bounds_check ; RUST NAME: core::panicking::panic_bounds_check
.text:0000000000005472                 mov     rsi, [rsp+0B8h+index] ; index
.text:0000000000005477                 mov     rdx, [rsp+0B8h+len] ; len
.text:000000000000547C                 call    rax ; core__panicking__panic_bounds_check ; RUST NAME: core::panicking::panic_bounds_check
.text:000000000000547E ; ---------------------------------------------------------------------------
.text:000000000000547E                 ud2
.text:0000000000005480 ; ---------------------------------------------------------------------------
.text:0000000000005480
.text:0000000000005480 loc_5480:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+159↑j
.text:0000000000005480                 lea     rdi, stru_382E8 ; location
.text:0000000000005487                 lea     rax, core__panicking__panic_bounds_check ; RUST NAME: core::panicking::panic_bounds_check
.text:000000000000548E                 mov     rsi, [rsp+0B8h+b] ; index
.text:0000000000005493                 mov     rdx, [rsp+0B8h+len] ; len
.text:0000000000005498                 call    rax ; core__panicking__panic_bounds_check ; RUST NAME: core::panicking::panic_bounds_check
.text:000000000000549A ; ---------------------------------------------------------------------------
.text:000000000000549A                 ud2
.text:000000000000549C ; ---------------------------------------------------------------------------
.text:000000000000549C
.text:000000000000549C loc_549C:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+19F↑j
.text:000000000000549C                 lea     rdi, str_0      ; expr
.text:00000000000054A3                 lea     rdx, off_38300  ; "rust.rs"
.text:00000000000054AA                 lea     rax, core__panicking__panic ; RUST NAME: core::panicking::panic
.text:00000000000054B1                 mov     esi, 21h ; '!'  ; expr
.text:00000000000054B6                 call    rax ; core__panicking__panic ; RUST NAME: core::panicking::panic
.text:00000000000054B8 ; ---------------------------------------------------------------------------
.text:00000000000054B8                 ud2
.text:00000000000054BA ; ---------------------------------------------------------------------------
.text:00000000000054BA
.text:00000000000054BA loc_54BA:                               ; CODE XREF: rust::bubble_sort::h0777bc845caabc60+C5↑j
.text:00000000000054BA                 lea     rdi, str_0      ; expr
.text:00000000000054C1                 lea     rdx, off_38318  ; "rust.rs"
.text:00000000000054C8                 lea     rax, core__panicking__panic ; RUST NAME: core::panicking::panic
.text:00000000000054CF                 mov     esi, 21h ; '!'  ; expr
.text:00000000000054D4                 call    rax ; core__panicking__panic ; RUST NAME: core::panicking::panic
.text:00000000000054D4 _ZN4rust11bubble_sort17h0777bc845caabc60E endp
.text:00000000000054D4
.text:00000000000054D6 ; ---------------------------------------------------------------------------
.text:00000000000054D6                 ud2
.text:00000000000054D6 ; } // starts at 5270
.text:00000000000054D6 ; ---------------------------------------------------------------------------
.text:00000000000054D8                 align 20h
```

**Comment by XVilka** [comment]
And this is what latest Binary Ninja shows:
```
rust::bubble_sort::h0777bc845caabc60:
00005270  4881ecb8000000     sub     rsp, 0xb8
00005277  4889bc2490000000   mov     qword [rsp+0x90 {var_28}], rdi
0000527f  4889b42498000000   mov     qword [rsp+0x98 {var_20}], rsi
00005287  48897c2448         mov     qword [rsp+0x48 {var_70}], rdi
0000528c  4889742440         mov     qword [rsp+0x40 {var_78}], rsi
00005291  e8ea090000         call    core::slice::_$LT$impl$u...5b$T$u5d$$GT$::len::hbaf7153778228b73
00005296  4889442450         mov     qword [rsp+0x50 {var_68}], rax
0000529b  c644245f01         mov     byte [rsp+0x5f {var_59}], 0x1

000052a0  f644245f01         test    byte [rsp+0x5f {var_59}], 0x1
000052a5  7508               jne     0x52af

000052a7  4881c4b8000000     add     rsp, 0xb8
000052ae  c3                 retn     {__return_addr}

000052af  c644245f00         mov     byte [rsp+0x5f {var_59}], 0x0
000052b4  488b442450         mov     rax, qword [rsp+0x50 {var_68}]
000052b9  48c7442460010000…  mov     qword [rsp+0x60 {var_58}], 0x1
000052c2  4889442468         mov     qword [rsp+0x68 {var_50_1}], rax
000052c7  488b7c2460         mov     rdi, qword [rsp+0x60 {var_58}]  {0x1}
000052cc  488b742468         mov     rsi, qword [rsp+0x68 {var_50_1}]
000052d1  e83a140000         call    _$LT$I$u20$as$u20$core.....tor$GT$::into_iter::h12f1e7995fb38cf5
000052d6  4889442438         mov     qword [rsp+0x38 {var_80_1}], rax
000052db  4889542430         mov     qword [rsp+0x30 {var_88_1}], rdx
000052e0  488b442438         mov     rax, qword [rsp+0x38 {var_80_1}]
000052e5  4889442470         mov     qword [rsp+0x70 {var_48}], rax
000052ea  488b4c2430         mov     rcx, qword [rsp+0x30 {var_88_1}]
000052ef  48894c2478         mov     qword [rsp+0x78 {var_40_1}], rcx

000052f4  488d7c2470         lea     rdi, [rsp+0x70 {var_48}]
000052f9  e812130000         call    core::iter::range::_$LT$...LT$A$GT$$GT$::next::h707e8283b20ce50a
000052fe  4889942488000000   mov     qword [rsp+0x88 {var_30_1}], rdx
00005306  4889842480000000   mov     qword [rsp+0x80 {var_38_1}], rax
0000530e  488b842480000000   mov     rax, qword [rsp+0x80 {var_38_1}]
00005316  4885c0             test    rax, rax
00005319  7404               je      0x531f

0000531b  eb00               jmp     0x531d

0000531d  eb23               jmp     0x5342

0000531f  488b442450         mov     rax, qword [rsp+0x50 {var_68}]
00005324  482d01000000       sub     rax, 0x1
0000532a  0f92c1             setb    cl
0000532d  f6c101             test    cl, 0x1
00005330  4889442428         mov     qword [rsp+0x28 {var_90_1}], rax
00005335  0f857f010000       jne     0x54ba

0000533b  e9f7000000         jmp     0x5437

00005340  0f 0b                                                                                            ..

00005342  488b842488000000   mov     rax, qword [rsp+0x88 {var_30_1}]
0000534a  48898424a0000000   mov     qword [rsp+0xa0 {var_18_1}], rax
00005352  48898424a8000000   mov     qword [rsp+0xa8 {var_10_1}], rax
0000535a  48898424b0000000   mov     qword [rsp+0xb0 {var_8_1}], rax
00005362  4889c1             mov     rcx, rax
00005365  4881e901000000     sub     rcx, 0x1
0000536c  0f92c2             setb    dl
0000536f  f6c201             test    dl, 0x1
00005372  4889442420         mov     qword [rsp+0x20 {var_98_1}], rax
00005377  48894c2418         mov     qword [rsp+0x18 {var_a0_1}], rcx
0000537c  0f85c4000000       jne     0x5446

00005382  488b442418         mov     rax, qword [rsp+0x18 {var_a0_1}]
00005387  488b4c2440         mov     rcx, qword [rsp+0x40 {var_78}]
0000538c  4839c8             cmp     rax, rcx
0000538f  0f92c2             setb    dl
00005392  f6c201             test    dl, 0x1
00005395  7505               jne     0x539c

00005397  e9c8000000         jmp     0x5464

0000539c  488b442418         mov     rax, qword [rsp+0x18 {var_a0_1}]
000053a1  48c1e002           shl     rax, 0x2
000053a5  488b4c2448         mov     rcx, qword [rsp+0x48 {var_70}]
000053aa  4801c1             add     rcx, rax
000053ad  488b442420         mov     rax, qword [rsp+0x20 {var_98_1}]
000053b2  488b542440         mov     rdx, qword [rsp+0x40 {var_78}]
000053b7  4839d0             cmp     rax, rdx
000053ba  400f92c6           setb    sil
000053be  40f6c601           test    sil, 0x1
000053c2  48894c2410         mov     qword [rsp+0x10 {var_a8_1}], rcx
000053c7  7505               jne     0x53ce

000053c9  e9b2000000         jmp     0x5480

000053ce  488b442420         mov     rax, qword [rsp+0x20 {var_98_1}]
000053d3  48c1e002           shl     rax, 0x2
000053d7  488b4c2448         mov     rcx, qword [rsp+0x48 {var_70}]
000053dc  4801c1             add     rcx, rax
000053df  488b7c2410         mov     rdi, qword [rsp+0x10 {var_a8_1}]
000053e4  4889ce             mov     rsi, rcx
000053e7  e894140000         call    core::cmp::impls::_$LT$i...or$u20$i32$GT$::gt::hcb39ce7110420639
000053ec  8844240f           mov     byte [rsp+0xf {var_a9_1}], al
000053f0  8a44240f           mov     al, byte [rsp+0xf {var_a9_1}]
000053f4  a801               test    al, 0x1
000053f6  7502               jne     0x53fa

000053f8  eb38               jmp     0x5432

000053fa  488b442420         mov     rax, qword [rsp+0x20 {var_98_1}]
000053ff  482d01000000       sub     rax, 0x1
00005405  0f92c1             setb    cl
00005408  f6c101             test    cl, 0x1
0000540b  48890424           mov     qword [rsp {var_b8_1}], rax
0000540f  0f8587000000       jne     0x549c

00005415  488b7c2448         mov     rdi, qword [rsp+0x48 {var_70}]
0000541a  488b742440         mov     rsi, qword [rsp+0x40 {var_78}]
0000541f  488b1424           mov     rdx, qword [rsp {var_b8_1}]
00005423  488b4c2420         mov     rcx, qword [rsp+0x20 {var_98_1}]
00005428  e8b30a0000         call    core::slice::_$LT$impl$u...b$T$u5d$$GT$::swap::h5ca623ef598b2bb1
0000542d  c644245f01         mov     byte [rsp+0x5f {var_59}], 0x1

00005432  e9bdfeffff         jmp     0x52f4

00005437  488b442428         mov     rax, qword [rsp+0x28 {var_90_1}]
0000543c  4889442450         mov     qword [rsp+0x50 {var_68}], rax
00005441  e95afeffff         jmp     0x52a0

00005446  488d3dc34b0200     lea     rdi, [rel str.0]  {"attempt to subtract with overflo…"}
0000544d  488d15642e0300     lea     rdx, [rel data_382b8]
00005454  488d05d5040200     lea     rax, [rel core::panicking::panic::h871cd04a3fc0d7bc]
0000545b  be21000000         mov     esi, 0x21
00005460  ffd0               call    rax  {core::panicking::panic::h871cd04a3fc0d7bc}
{ Does not return }
```

**Comment by XVilka** [comment]
After digging it turned out that the old code wasn't working in this case, but you can see this `values` is of `struct._mut__i32_` type which has two members - and the length that overlaps with that `var_98h` (it's `values.len` in fact):
```
[0x00005270]> tk~mut__i32
_mut__i32_=struct
struct._mut__i32_=data_ptr,length
struct._mut__i32_.data_ptr=i32 *,0,0
struct._mut__i32_.length=usize,8,0
```
Moreover, this is missing part in analysis - it should put `values.len` instead of `rsp + 0x98h` afterwards, but that's a separate issue.
Same with iter:
```
[0x00005270]> tk~Range_usize_
Option_core::ops::range::Range_usize__=struct
Option_core::option::Option_core::ops::range::Range_usize___=struct
Range_usize_=struct
Result_core::option::Option_core::ops::range::Range_usize____std::thread::local::AccessError_=struct
Rev_core::ops::range::Range_usize__=struct
func.sym.core::str::traits::__impl_core::slice::SliceIndex_str__for_core::ops::range::Range_usize__::index::__u7b__u7b_closure_u7d__u7d_::h2155fee845c05e30.arg.0=int64_t,arg1
func.sym.core::str::traits::__impl_core::slice::SliceIndex_str__for_core::ops::range::Range_usize__::index::__u7b__u7b_closure_u7d__u7d_::h2155fee845c05e30.args=1
func.sym.core::str::traits::__impl_core::slice::SliceIndex_str__for_core::ops::range::Range_usize__::index::__u7b__u7b_closure_u7d__u7d_::hd8c28b300f2d3170.arg.0=uint32_t,arg1
func.sym.core::str::traits::__impl_core::slice::SliceIndex_str__for_core::ops::range::Range_usize__::index::__u7b__u7b_closure_u7d__u7d_::hd8c28b300f2d3170.args=1
struct.Range_usize_=start,end
struct.Range_usize_.end=usize,8,0
struct.Range_usize_.start=usize,0,0
struct.Rev_core::ops::range::Range_usize__=iter
struct.Rev_core::ops::range::Range_usize__.iter=Range<usize>,0,0
```

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**